File: Medium/186. Reverse Words in a String II/solution.java
class Solution {
    public void reverseWords(char[] s) {
        String str = new String(s);
        String[] words = str.split("\\s+");
        
        int left = 0;
        int right = words.length - 1;
        while (left < right) {
            String tmp = words[left];
            words[left] = words[right];
            words[right] = tmp;
            left++;
            right--;
        }
        
        int index = 0;
        for (int i = 0; i < words.length; i++) {
            for (char c: words[i].toCharArray()) {
                s[index++] = c;
            }
            if (i != words.length - 1) {
                s[index++] = ' ';
            }
        }
        return;
    }
}

/**
 * solution 2
 */
class Solution {
    public void reverseWords(char[] s) {
        reverse(s, 0, s.length - 1);
        
        int start = 0;
        for (int i = 0; i < s.length; i++) {
            if (s[i] == ' ') {
                reverse(s, start, i - 1);
                start = i + 1;
            }
        }
        
        reverse(s, start, s.length - 1);
        return;
    }
    
    private void reverse(char[] s, int left, int right) {
        
        while (left < right) {
            char c = s[left];
            s[left] = s[right];
            s[right] = c;
            left++;
            right--;
        }
    }
}

File: Medium/186. Reverse Words in a String II/solution.py
class Solution:
    def reverseWords(self, s: List[str]) -> None:
        """
        Do not return anything, modify s in-place instead.
        """
        left: int = 0
        words = []
        for right in range(len(s)):
            if s[right] == " ":
                word = "".join(s[left: right])
                words.append(word)
                left = right + 1

        word = "".join(s[left:right + 1])
        words.append(word)
        print(words)
        idx: int = 0
        for indice in range(len(words) - 1, 0, -1):
            word = words[indice]
            for char in word:
                s[idx] = char
                idx += 1
            
            s[idx] = " "
            idx += 1
        
        word = words[0]
        for char in word:
            s[idx] = char
            idx += 1
        
        return
    